Include/internal/pycore_object.h

```diff
@@ -655,24 +656,73 @@ _PyDictOrValues_IsValues(PyDictOrValues dorv)
     return ((uintptr_t)dorv.values) & 1;
 }
 
+// Should only be used when we know the object isn't racing with other
+// threads (e.g. finalization or GC).
 static inline PyDictValues *
 _PyDictOrValues_GetValues(PyDictOrValues dorv)
 {
     assert(_PyDictOrValues_IsValues(dorv));
     return (PyDictValues *)(dorv.values + 1);
 }
 
+static inline PyDictValues *
+_PyDictOrValues_TryGetValues(PyObject *obj)
```

Comment:
I think this function needs to be split up for read vs. write cases. The current logic doesn't seem quite right to me.

For readers, we generally want to fall back to a slower locking code path if it's not owned by the current thread and not marked shared.

For writers -- or really any case that hold's the lock -- we don't want a nested critical section. We also don't require that the object be marked as shared, although writers may mark it as shared as an optimization (so that future readers from non-owning threads don't need to acquire the lock).

Reply:
Ahh yes, I'm missing locking on writes entirely...  I was thinking maybe we wouldn't need that for the shared values but that's not right.

I don't think we want to fall back to a slower locked version for reads here, mainly because it complicates the changes down stream in a similar way to making the locks for writes be present.  We'd need to have something like a TryGetValue instead of TryGetValues and then all of the consumers would need to pass in the index.  Instead we just need to lock to make the object shared and we can re-run the unlocked version safely.

Reply:
> I don't think we want to fall back to a slower locked version for reads here, mainly because it complicates the changes down stream in a similar way to making the locks for writes be present.

That seems reasonable. I'm a bit worried that the more complex structure of `_PyDictOrValues_TryGetValues` will prevent efficient inlining, but we can address that later.